KEYCLOAK-18341 Support JWKS OAuth2 Client Metadata in the "by value" key loading method

y-tabata opened the pull request:
JIRA: https://issues.redhat.com/browse/KEYCLOAK-18341

# Changes

- Now we can register the client app's public keys by "jwks" OAuth2 Client Metadata via Admin Console (Admin REST API) as realm attribute parameter.

- New components in the "Keys" tab.
  - `Use JWKS` switch
  - `JWKS` text area
    - Display this component when `Use JWKS` switch is ON

y-tabata committed:
KEYCLOAK-18341 Support JWKS OAuth2 Client Metadata in the "by value" key loading method

y-tabata commented:
@mposolda
This is the second step of "Improve support for setting keys for OIDC clients (https://issues.redhat.com/browse/KEYCLOAK-10462)" after https://github.com/keycloak/keycloak/pull/7874.
I would appreciate it if you review this.

y-tabata commented:
Docs PR: https://github.com/keycloak/keycloak-documentation/pull/1195

mposolda commented on themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:
```diff
@@ -471,6 +471,10 @@ use-jwks-url=Use JWKS URL
 jwks-url=JWKS URL
 jwks-url.tooltip=URL where client keys in JWK format are stored. See JWK specification for more details. If you use Keycloak client adapter with "jwt" credential, you can use URL of your app with '/k_jwks' suffix. For example 'http://www.myhost.com/myapp/k_jwks' .
+use-jwks=Use JWKS
+use-jwks.tooltip=If the switch is on, client public keys will be configurable in JWKS.
```
Please add note to this tooltip, that switch "Use JWKS" is mutually exclusive with the switch "Use JWKS URL". Also please update the tooltip for switch "Use JWKS URL" in same manner.

y-tabata replied:
I added.

mposolda reviewed (changes requested):
@y-tabata Thanks a lot for the PR and sorry it took longer for the review. I've added one inline comment regarding tooltips. In addition to that:

- Is it possible to rebase on top of latest Keycloak master?

- In relation to the tooltip note above, is it possible to add the validation on server-side, that when you try to update client and both "Use JWKS" and "Use JWKS URL" are ON, the error will be thrown? We have this validation for OIDC dynamic client registration only (in `DescriptionConverter.setPublicKey`), but will be nice to have for admin REST API too in my opinion.

- One related note is, that `DescriptionConverter.setPublicKey` supports to push only single key and only for the SIG use for the case when "jwks" was used. Will it makes sense to improve this to push the keys instead to the "jwks" attribute of the client?

y-tabata commented:
@mposolda Thanks for the review.

> Is it possible to rebase on top of latest Keycloak master?

I rebased.

> In relation to the tooltip note above, is it possible to add the validation on server-side, that when you try to update client and both "Use JWKS" and "Use JWKS URL" are ON, the error will be thrown? We have this validation for OIDC dynamic client registration only (in DescriptionConverter.setPublicKey), but will be nice to have for admin REST API too in my opinion.

I added a server-side check in `DefaultClientValidationProvider`.

> One related note is, that DescriptionConverter.setPublicKey supports to push only single key and only for the SIG use for the case when "jwks" was used. Will it makes sense to improve this to push the keys instead to the "jwks" attribute of the client?

Considering backward compatibility, I added "jwks_string" newly we can push keys not only a single SIG use key.

y-tabata commented:
@mposolda Could you review this?

tnorimat commented:
@mposolda If registering RSASSA-PSS signature algorithm's client public key is beneficial for supporting Open Banking Brasil (it only allows PS256), not via jwks_url, this PR is useful. Could you consider this point?

mposolda commented on services/src/main/java/org/keycloak/services/clientregistration/oidc/DescriptionConverter.java:
```diff
@@ -260,12 +264,19 @@ private static boolean setPublicKey(OIDCClientRepresentation clientOIDC, ClientR
+            configWrapper.setUseJwksString(false);
+            return true;
+        } else {
+            configWrapper.setUseJwksString(true);
```
My vote is to change whole if/else block to something like this (Not 100% sure it is syntactically correct, but I hope it will help to illustrate what I mean). In other words remove whole first block of this if/else block and in case that OIDC representation has "jwks" set, then just set "useJwksString" to true in Keycloak representation. I think this will allow to remove `useJwksString` field from OIDCClientRepresentation as pointed above.
```
        if (clientOIDC.getJwks() != null) {
           configWrapper.setUseJwksString(true);
            configWrapper.setJwksString(clientOIDC.getJwksString());
            configWrapper.setUseJwksUrl(false);
            return true;
            }
        } else {
            configWrapper.setUseJwksUrl(true);
            configWrapper.setJwksUrl(clientOIDC.getJwksUri());
            return true;
        }
```

y-tabata replied:
@mposolda I changed the whole if/else block. Is my understanding correct?

mposolda reviewed (changes requested):
@y-tabata @tnorimat Sorry for late response. Added review comments, please check if it makes sense to you. Thanks

mposolda commented:
@y-tabata Also rebase is needed to avoid conflict.

y-tabata commented on core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java:
```diff
@@ -62,6 +62,8 @@
 
     private String jwks_uri;
 
+    private String jwks_string;
```
@mposolda I see. I deleted `jwks_string` from OIDCClientRepresentation.

mposolda reviewed (changes requested):
@y-tabata Thanks for the updates!

One last small thing: I noticed that `DescriptionConverter.toExternalResponse` may worth update as well? For the case when "jwksString" is used, it can be good to update OIDCClientRepresentation with the "jwks" from Keycloak client representation?

Besides that, I am trying to run the internal pipeline to doublecheck there are no regressions in admin console tests. But I hope everything will be ok.

Thanks for your work on this!

y-tabata commented:
Thanks for your review.
> One last small thing: I noticed that `DescriptionConverter.toExternalResponse` may worth update as well? For the case when "jwksString" is used, it can be good to update OIDCClientRepresentation with the "jwks" from Keycloak client representation?

I changed it. Could you confirm again?

mposolda reviewed (approved):
@y-tabata Thanks! The pipeline run did not find any regression of your changes.

I am planning to merge the PR once we have approve also for the docs PR from Andy (our documentation guru).
